src/chunk.c

```diff
@@ -1210,13 +1212,62 @@ chunk_create_from_hypercube_after_lock(const Hypertable *ht, Hypercube *cube,
 													  prefix,
 													  get_next_chunk_id());
 
+	/* If we have compressoin enabled, we create the compressed chunk here
+	 * to reduce locking contention since we already use heavy locks to attach
+	 * chunks to the hypertable.
+	 */
+	if (ht->fd.compressed_hypertable_id != 0)
+	{
+		chunk->fd.compressed_chunk_id =
+			create_compressed_chunk(ht->fd.compressed_hypertable_id, schema_name);
+	}
+
 	chunk_add_constraints(chunk);
 	chunk_insert_into_metadata_after_lock(chunk);
 	chunk_create_table_constraints(ht, chunk);
 
 	return chunk;
 }
 
+static int32
+create_compressed_chunk(int32 compressed_hypertable_id, const char *schema_name)
+{
+	Hypertable *compress_ht = ts_hypertable_get_by_id(compressed_hypertable_id);
+	int32 chunk_id = get_next_chunk_id();
+	NameData *compress_chunk_table_name = palloc0(sizeof(*compress_chunk_table_name));
+
+	int namelen = snprintf(NameStr(*compress_chunk_table_name),
+						   NAMEDATALEN,
+						   "compress%s_%d_chunk",
+						   NameStr(compress_ht->fd.associated_table_prefix),
+						   chunk_id);
+
+	if (namelen >= NAMEDATALEN)
+		ereport(ERROR,
+				(errcode(ERRCODE_INTERNAL_ERROR),
+				 errmsg("invalid name \"%s\" for compressed chunk",
+						NameStr(*compress_chunk_table_name)),
+				 errdetail("The associated table prefix is too long.")));
+	Hypercube *c_cube = ts_hypercube_alloc(0);
+
+	Chunk *compressed_chunk =
+		chunk_create_only_table_after_lock(compress_ht,
+										   c_cube,
+										   schema_name,
+										   NameStr(*compress_chunk_table_name),
+										   NULL,
+										   chunk_id);
+
+	compressed_chunk->constraints = ts_chunk_constraints_alloc(1, CurrentMemoryContext);
+
+	chunk_add_constraints(compressed_chunk);
+	chunk_insert_into_metadata_after_lock(compressed_chunk);
+	chunk_create_table_constraints(compress_ht, compressed_chunk);
```

Comment:
These seem to be done both inside `create_compressed_chunk` and at one of the call sites. Do you need both? Is there a risk of constraints being added twice or metadata being added twice for the same chunk?